doc/lpc18/results.tex

```diff
@@ -10,24 +10,94 @@ \subsection{Testbed}
+For each test program, we use the following
+command from iproute2 tool to load it into kernel:
+\begin{verbatim}
+ip link set dev eth0 xdp obj xdp1.o verb
+\end{verbatim}
+The Intel 10GbE X540 NIC is running ixgbe driver with 16 RX queues
+set-up. Since the source server is sending single UDP flow, packets
+always arrive at a single ring ID.  As a result, we collect the number
+of packets being dropped at this ring.
+
+\subsection{Results}
+To compare the performance, we first started by manually writing two
+XDP program. The first one simply does nothing but drop all packets by
+returning XDP\_DROP. The second program also does nothing but returns
+XDP\_TX, which forwards the packet to the receiving port.  Both programs
+consists of only two BPF instructions.
+0: (b7) r0 = 1
+1: (95) exit
+
+Then We attached the following P4 programs to the device receiving the
+rate of 14.88~Mpps to evaluate the overhead introduce by the P4C-XDP
+compiler.
+\begin{itemize}
+\item xdp1.p4: Parse Ethernet/IPv4 header, deparse it, and drop.
+\item xdp3.p4: Parse Ethernet/IPv4 header, lookup an mac address
+table, deparse it, and drop.
+\item xdp6.p4: Parse Ethernet/IPv4 header, lookup and get an 
+ttl from metadata, set to IPv4 header, deparse it, and drop.
+\item xdp7.p4: Parse Ethernet/IPv4/UDP header, modify source port
+and source IP, recalculate checksum, deparse, and drop.
+\item xdp11.p4: Parse Ethernet/IPv4 header, swap src/dst mac address,
+deparse it, and send back to the same port (XDP\_TX).
+\item xdp15.p4: Parse Ethernet header, insert a customized 8-byte header,
+deparse it, and send back to the same port (XDP\_TX).
+\end{itemize}
+
+
+As show in Table X, the xdp1.p4 shows the baseline overhead introduced
+by adding the parser and deparser, dropping the rate from 14.4~Mpps to
+8.1~Mpps. The xdp3.p4 drops another million packet per second due to
+calling the eBPF map lookup function to do a lookup.
+The xdp6.p4 shows significant overhead because it (why?) 
+Surprisingly, the xdp7.p4 does extra parsing to the UDP header and
+checksum recalculation, but the overhead is moderate.
+
+Finally, the xdp11.p4 and xdp15.p4 tests the XDP\_TX performance.
+xdp15.p4 involves the bpf\_adjust\_head helper function.
+Interestingly, it does not incur much over head because there is
+already a reserved space in front of every packet frame.
 
 \begin{table}
 \centering
 \small
 \begin{tabular}{lll}
-  \underline{P4 program} & \underline{Description} & \underline{Performance} \\
-  xdp1.p4 & parse L2/L3, drop if non-IP packet &  Mpps \\
-  xdp2.p4 & parse L2/L3/L4, drop if non-IP packet &  Mpps \\
-  xdp3.p4 & parse L2/L3, lookup &  Mpps \\
+  \underline{P4 program} & \underline{CPU Util.} & \underline{Mpps} \\
+  SimpleDROP & 75\% & 14.4 \\
+  SimpleTX & WIP & WIP \\
+  xdp1.p4 &  100\% &  8.1 \\
+  xdp3.p4 &  100\% &  7.1 \\
+  xdp6.p4 &  100\% &  2.5 \\
+  xdp7.p4 &  100\% &  5.7 \\
+  xdp11.p4 &  100\% &  4.7 \\
+  xdp15.p4 &  100\% &  5.5 \\
```

Comment:
Apologies if you do mention this somewhere and I missed it, but it would be good to explicitly say whether 'CPU Util.=100%' means 100% of one CPU core on a multi-core machine, or 100% of all CPU cores on a multi-core machine, and in that case, how many cores there are.

Reply:
Also, what model and clock speed of CPU was used.

Reply:
Thanks a lot for the feedback, I will update in next version!